tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -2479,20 +2493,23 @@ void HWCommandQueue::enqueue_program(Program& program, bool blocking) {
     TT_FATAL(sub_device_ids.size() == 1, "Programs must be executed on a single sub-device");
     if (not program.is_finalized()) {
         program.finalize(device);
-        TT_FATAL(!this->manager.get_bypass_mode(), "Tracing should only be used when programs have been cached");
-        if (const auto &kernels_buffer = program.get_kernels_buffer()) {
-            // Only stall for used sub-devices
+    }
+    if (program.binaries_on_device_status(device->id()) == ProgramBinaryStatus::NotSent) {
+        // Write program binaries to device
+        program.allocate_kernel_bin_buf_on_device(device);
+        if (program.get_program_transfer_info().binary_data.size()) {
             this->enqueue_write_buffer(
-                *kernels_buffer, program.get_program_transfer_info().binary_data.data(), false, sub_device_ids);
+                *program.get_kernels_buffer(device), program.get_program_transfer_info().binary_data.data(), false, sub_device_ids);
```

Comment:
Can these operations fail? Can the program API be more like "compare and swap", returning a boolean indicating the status? This would prevent the need for manually setting the status on line 2504.

Reply:
I'm not entirely sure what this means. Are you asking about `enqueue_write_buffer`? This is a generic API to move data from host to device, it wouldn't really make sense to return a program specific status here.

Reply:
I'm referring to updating program binary status in lock-step. This resembles a transaction: (1) read status, (2) perform some action based on the read, (3) write back the status. My question is basically if it is possible for (2) to fail, and if yes how to act in step (3) in this case? Regarding compare-and-swap - my thinking here was basically if you forget to do (3) and still do (2). What happens then?

Is this code going to be executed concurrently, now or in the future?

These scenarios might not apply, btw, if we use this functionality in a very isolated way. Let me know!

Reply:
I think I understand what you mean now. The suggested approach is cleaner and less error prone for reasons you mentioned. One way to do this without relying on a status from `enqueue_write_buffer` would be to encapsulate the writing of the binaries + the status update inside a helper function in the program class.
The only reason I'm not comfortable doing this just yet is because this status is modified again by the `CommandQueue` object downstream in `assemble_device_commands` (having the program object do this internally would be unclean for now). We would thus have an inconsistency where `NotSent` -> `InFlight` is modified by the program and `InFlight` -> `Committed` is modified by the Command Queue.

I think this will get cleaned up in future, as part of a more modular program class. I'll be working on parts of this cleanup due to code reuse between `Program` and `MeshWorkload`.